Browser side datatables rendering (Fix: #75, Fix: #76, Fix #77, Fix #80, Fix #90, Fix #91, Fix #94, Fix: #99, Fix: #101, Fix #102)

timriker opened the pull request:
This is ready for merging. Rendering times on our setup drop from 5 minutes to about 15 seconds. Build times are longer as the code looks through iRules and datagroups for pool references and now includes linking information in the json files.

- version 5.1.7
- console menu on main screen
- iRules tab include all iRules
- new datatables tab
- new pools tab (orphans are here)
- tabs use datatables and render on demand, not on load
- ErrorReportAnyway added to generate reports when there are errors
- virtual servers include client and server ssl profiles
- ASM types are included read from `asmpolicies.json`
- made the logo transparent (unrelated)
- relative paths for resource loading from javascript
- use Map() for pool lookups to speed up loading
- log stats at the end of the build, and when loading
- more adcLinks in various places, links open in new window/tab
- server side processing to build irule referenced pools and datagroups
- server side processing to build datagroup pool references
- new MaxPools to limit status checks for too many open pools

I recommend looking at the diffs with whitespace ignored. Add &w=1 to the diff url, like:
https://github.com/epacke/BigIPReport/pull/93/files?diff=unified&w=1

epacke commented on Move the content of this folder to the configured report root/js/jquery.min.js:
```diff
@@ -1,4 +1,4 @@
-/*! jQuery v2.2.1 | (c) jQuery Foundation | jquery.org/license */
-!function(a,b){"object"==typeof module&&"object"==typeof module.exports?module.exports=a.document?b(a,!0):function(a){if(!a.document)throw new Error("jQuery requires a window with a document");return b(a)}:b(a)}("undefined"!=typeof window?window:this,function(a,b){var c=[],d=a.document,e=c.slice,f=c.concat,g=c.push,h=c.indexOf,i={},j=i.toString,k=i.hasOwnProperty,l={},m="2.2.1",n=function(a,b){re ...
-}else c=void 0;return c}n.extend({hasData:function(a){return O.hasData(a)||N.hasData(a)},data:function(a,b,c){return O.access(a,b,c)},removeData:function(a,b){O.remove(a,b)},_data:function(a,b,c){return N.access(a,b,c)},_removeData:function(a,b){N.remove(a,b)}}),n.fn.extend({data:function(a,b){var c,d,e,f=this[0],g=f&&f.attributes;if(void 0===a){if(this.length&&(e=O.get(f),1===f.nodeType&&!N.get( ...
-e=n.propHooks[b]),void 0!==c?e&&"set"in e&&void 0!==(d=e.set(a,c,b))?d:a[b]=c:e&&"get"in e&&null!==(d=e.get(a,b))?d:a[b]},propHooks:{tabIndex:{get:function(a){var b=n.find.attr(a,"tabindex");return b?parseInt(b,10):cb.test(a.nodeName)||db.test(a.nodeName)&&a.href?0:-1}}},propFix:{"for":"htmlFor","class":"className"}}),l.optSelected||(n.propHooks.selected={get:function(a){var b=a.parentNode;return ...
```
Why was this file changed?

timriker replied:
lineendings. I set .js files to text in .gitattribites

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
Using "on" else where. Change this to "on("focus", function () {" ?

epacke replied:
Probably old sins from my side. Hard to tell with this diff.

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
How about adding some white-space before and after "=", ";"?

epacke replied:
How about renaming this GetPoolMemberStatus? I find it easier to understand what the function does if a verb is present (a bit "PowerShelly").

epacke replied:
How about GetVirtualServerStatus?

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
Curious. Why not  use href here?

timriker replied:
Your comments are not getting attached to the right line numbers. Is this about the loadbalancer column? Just preference. I didn't want the default href rendering, and I wanted links to open in a new window/tab. Easy to change back if you like.

epacke replied:
Default href rendering as in the same window? Could use target for that. Either way, it's just different ways to skin a cat. Just curious.

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
Double new lines here

timriker replied:
again. not getting attached to the right line numbers. I see this on bigipreport.js line 1818. I don't see double new lines there.

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
Double new lines

epacke replied:
(These are old sins from me, it's just more clear on github due to their long line spacing)

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
mstatus -> mStatus?

timriker replied:
ok. vsstatus too I expect? :)

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
How about vsStatus?

timriker replied:
:)

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
cellData and colIndex are never used.

timriker replied:
Yep, just default parameters from datatables.

timriker commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2541 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
wish I could see what lines you are commenting on... :(

epacke commented:
I'll hold off with reviewing the PR until you're done. Otherwise I'll have to re-do it again. :)

epacke commented:
That's great! It's quite a big change so chances are that I won't be able to make it in a single session (because of the kids). I'll hold off for a while in case the "beta testers" finds some things.

Amazing work so far, feels like the missing puzzle piece has been found!

epacke commented:
Suggestion: Branch the current state and add features there. Then I can review it. Is all the functionality from the old report present in this release?

timriker commented:
I should have branched before the last couple commits. The last two commits before this message moved the console menu to the main page, and converted the iRules datatable to showing all loaded iRules.

Before that change, everything worked as before except: there are edit and details links in more places than before, pools have status indicators, minor visual layout updates, etc.

epacke commented:
I've tested the latest version now. Wow, so many great ideas and new angles! However, I was under the impression that we had agreed to discuss any additional UX changes but since then it seems like ie. a pool table, data grouplist view has been added.

Is it possible to take a step back and focus on the core things, then discuss the potential additions and how to design? How about a Skype call? When are you available?

timriker commented:
Yes, I did add a couple new tabs. iRules (now all rules), Pools (with orphans) and Data Groups. I didn't consider adding the new tabs as a UX change, but I guess it is.

My current changes are to avoid setting up each tab until it is used. This cuts down the load times some more. The moving of code does make the pull request even harder to review. Sorry about that.

I'm planning to work on more discovery at build time. I'd like to identify "String" datagroups as URLs or virtual hosts, or pools. I'd like to include pools as a datagroup attribute, and track that back to virtual servers so pools show under virtualgroups even when the pool is references through vs->irule->datagroup->pool.

This is a busy weekend for me, so I'm not sure when we could skype.

Overall impressions? We're using BigIPReport to do a lot of cleanup here, so it's been very valuable!

I've experimented with making the iControl SOAP calls directly in powershell without using the iControl Snapin. The portable powershell does not include Snapin support. The soap calls are not as readable in my current testing, but there's not a lot of code to replace. Perhaps I'll open a ticket and paste an example there.

timriker commented:
I'll pick a place and branch after that to make the review easier. Be sure to use the w=1 url when looking at my diffs as otherwise the whitespace cleanups hide any useful changes.

My latest changes I added things like setupVirtualServersTable() and moved all the setup for bigipTable into that function. We currently still need to call that one on startup, but we don't need to call setupPoolTable() until the user actually picks the pools tab. It would be nice to call all the setup*() functions after initial page load so the tabs still load fast later. If the user comes in with mainsection=help on the url, in theory we can avoid loading the virtual servers tab until after the help page draws. In practice, I'm not quite there, as preferences tab assumes we have a bigipTable.
I moved bigipTable and the other tables under siteData. I guess we could have a siteData.tables.virtualsevers or similar to we can iterate over the tables and rerender them if needed.
So many options! :)

timriker commented:
I branched all my changes since the console move. The pull request at this point is more manageable. This gets us to client side rendering for the main table. Includes the asm policies json and NAT support. At this point the same functionality is there as before, with the client side speed improvements. There might be an Edge ES6 fix needed that's in my branch. Remember to review with the w=1 to avoid all the whitespace. This upgrades jQuery and Datatables. They are the upstream current releases at the point I downloaded them.

epacke commented on Move the content of this folder to the configured report root/js/bigipreport.js:
```diff
@@ -1,2385 +1,2595 @@
-						//Set an initial value of dg
-						var dg = ""
-						
-						//If the command is class. I've chosen n
```
Old sins from me, but let's declare this one with var?

timriker replied:
Your comments show on line 1818. Not sure which you want as a var.

epacke commented:
Really nice Tim, I don't have much to remark on at all. Excellent work!

I'd love to have a short session with you so you can explain the render process. I can't wrap my mind around it and it'd be nice to get an explanation and be able to ask questions.

Two things that I saw:
* The console is not even present in this release
* Loadbalancers.json is written as an object rather than as an array containing objects if only one LB is defined in the configuration file

timriker commented:
Can you confirm which commit you're testing on? Just do a git log and post the top hash. This commit:

[https://github.com/timriker/BigIPReport/commit/9176056c5fb25ce14ea9a8f2affd14db0a22eb50](https://github.com/timriker/BigIPReport/commit/9176056c5fb25ce14ea9a8f2affd14db0a22eb50)

which was included in test release 5.1.3 included the fix for the non-array state of loadbalancers.json.

The console menu/tabs promotion happened in

[https://github.com/timriker/BigIPReport/commit/728e691ecf536d8c6f1a902a30bc4fc291d1cf5d](https://github.com/timriker/BigIPReport/commit/728e691ecf536d8c6f1a902a30bc4fc291d1cf5d)

Which should only be included in my "promoteconsole" branch and NOT included in this pull request. I rewrote history on my repo to get to that state. I moved all my recent changes off of master and onto my promoteconsole branch. You should still see the console on this pull request. If you already had my repo checked out someplace, you will need to do something like this:
```
$ git checkout master
$ git reset --hard origin/master
```

My current branch is where release 5.1.5 lives.

https://github.com/timriker/BigIPReport/tree/promoteconsole

I'll send an email with some info on how the datatables rendering works. It's slick when it works, but sometimes hard to debug. The biggest thing that gets me is I forget to have rendering functions always return the rendered result. The errors when they return nothing are not very clear. Simple, stupid error to fix, but I'm surprised how many times I work through building the html and then forget to actually return it. :)
